Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm

```diff
@@ -1094,6 +1114,66 @@ static _WKWebExtensionError toAPI(WebExtension::Error error)
 #endif
 }
 
+bool WebExtension::hasOptionsPage()
+{
+    populatePagePropertiesIfNeeded();
+    return !!m_optionsPagePath;
+}
+
+bool WebExtension::hasOverrideNewTabPage()
+{
+    populatePagePropertiesIfNeeded();
+    return !!m_overrideNewTabPagePath;
+}
+
+NSString *WebExtension::optionsPagePath()
+{
+    populatePagePropertiesIfNeeded();
+    return m_optionsPagePath.get();
+}
+
+NSString *WebExtension::overrideNewTabPagePath()
+{
+    populatePagePropertiesIfNeeded();
+    return m_overrideNewTabPagePath.get();
+}
+
+void WebExtension::populatePagePropertiesIfNeeded()
+{
+    if (!manifestParsedSuccessfully())
+        return;
+
+    if (m_parsedManifestPageProperties)
+        return;
+
+    m_parsedManifestPageProperties = true;
```

Comment:
Nit: Should we hold off on claiming we successfully parsed until the parsing is done? There don't seem to be any early returns, so maybe this is okay as-is.

Reply:
Yeah, good point. I set it early to avoid any issues of reentrancy.